### ISSUE TITLE
Fix Docker networking

### ISSUE THREAD

**Issue by iainelder** [iainelder/dotfiles#8]
tl;dr: Reinstall Docker. Here's everything I tried.

---

On my machine the local testing script is broken. APT can't resolve `archive.ubuntu.com` or `security.ubuntu.com`.

```
$ ./test_local.bash
+ script=_
++ pwd
+ docker run -it --rm --volume /var/run/docker.sock:/var/run/docker.sock --mount type=bind,source=/home/isme/Repos/dotfiles,target=/code ubuntu:20.04 bash -c '/code/scripts/install_in_docker.bash _'
+ program=_
+++ dirname /code/scripts/install_in_docker.bash
++ cd /code/scripts
++ pwd
+ script_dir=/code/scripts
+ /code/scripts/prepare_docker.bash
+ apt-get --yes update --error-on=any
Err:1 http://archive.ubuntu.com/ubuntu focal InRelease                   
  Temporary failure resolving 'archive.ubuntu.com'
Err:2 http://security.ubuntu.com/ubuntu focal-security InRelease         
  Temporary failure resolving 'security.ubuntu.com'
Err:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease           
  Temporary failure resolving 'archive.ubuntu.com'
Err:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease
  Temporary failure resolving 'archive.ubuntu.com'
Reading package lists... Done        
E: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal/InRelease  Temporary failure resolving 'archive.ubuntu.com'
E: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal-updates/InRelease  Temporary failure resolving 'archive.ubuntu.com'
E: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal-backports/InRelease  Temporary failure resolving 'archive.ubuntu.com'
E: Failed to fetch http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease  Temporary failure resolving 'security.ubuntu.com'
E: Some index files failed to download. They have been ignored, or old ones used instead.
```

`apt-get update` works just fine outside Docker.

Google `Docker Ubuntu Temporary failure resolving 'archive.ubuntu.com'`.

[Faithful Anere](https://medium.com/@faithfulanere/solved-docker-build-could-not-resolve-archive-ubuntu-com-apt-get-fails-to-install-anything-9ea4dfdcdcf2) helps me diagnose the problem.

Failthful shows how to use the busybox container to run the nslookup command. The "no servers could be reached" message shows that DNS is broken for all Docker containers.

```console
$ docker run --rm busybox nslookup google.com
;; connection timed out; no servers could be reached
```

Faithful's "quick fix" uses the `docker run --dns` option. To get the option value he suggests using Ubuntu's NetworkManager CLI (new to me!) to get the address of the DNS server.

```console
$ nmcli dev show | grep 'IP4.DNS'
IP4.DNS[1]:                             192.168.1.1
```

For me using the `--dns` option with NetworkManager's DNS server gives the same result.

```console
$ docker run --dns 192.168.1.1 --rm busybox nslookup google.com
;; connection timed out; no servers could be reached
```

Faithful also states something that appears not to be true in my case: "Containers use google dns server 8.8.8.8 to resolve DNS".

Even when I use `--dns` to force Google DNS, the lookup fails.

```console
$ docker run --dns 8.8.8.8 --rm busybox nslookup google.com
;; connection timed out; no servers could be reached
```

Faithful's "system-wide fix" updates the [Docker daemon settings](https://docs.docker.com/reference/cli/dockerd/) (also new to me) in `/etc/docker/daemon.json`.

For me the file doesn't exist yet.

```console
$ cat /etc/docker/daemon.json
cat: /etc/docker/daemon.json: No such file or directory
```

Write Faithful's minimal override configuration.

```json
{
    "dns": ["192.168.1.1", "8.8.8.8"]
}
```

But nslookup in Docker still fails.

```
$ docker run --rm busybox nslookup google.com
;; connection timed out; no servers could be reached
```

Faithful credits a [Stack Overflow question](https://stackoverflow.com/questions/24991136/docker-build-could-not-resolve-archive-ubuntu-com-apt-get-fails-to-install-a) for these steps.

There [Gerald Hansen](https://stackoverflow.com/a/50413039/111424) shows that the `--network=host` option worked for him.

It works for me too.

```console
$ docker run --network=host --rm busybox nslookup google.com
Server:		127.0.0.53
Address:	127.0.0.53:53

Non-authoritative answer:
Name:	google.com
Address: 142.250.200.14

Non-authoritative answer:
Name:	google.com
Address: 2a00:1450:4009:826::200e
```

From `man docker-run`, the default network type is `bridge`, which connects the container to the default Docker bridge via veth interfaces.

Google `docker bridge network broken`.

[A thread in the Docker forum](https://forums.docker.com/t/docker-bridge-networking-does-not-work-in-ubuntu-22-04/136326) shows another way to diagnose the problem using the ping command. Ping doesn't work either in the bridge network.

Bridge network ping:

```console
$ docker run --rm busybox ping google.com
ping: bad address 'google.com'
```

Host network ping:

```console
$ docker run --rm --network=host busybox ping google.com
PING google.com (172.217.16.238): 56 data bytes
64 bytes from 172.217.16.238: seq=0 ttl=55 time=49.265 ms
64 bytes from 172.217.16.238: seq=1 ttl=55 time=41.997 ms
64 bytes from 172.217.16.238: seq=2 ttl=55 time=51.548 ms
64 bytes from 172.217.16.238: seq=3 ttl=55 time=47.185 ms
^C
--- google.com ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 41.997/47.498/51.548 ms
```

The thread discussions a solution for Ubuntu 22 using netplan. I'm using Ubuntu 20 and it doesn't apply to me because the netplan config file `/etc/netplan/50-cloud-init.yaml` doesn't exist.

https://superuser.com/questions/1745422/docker-containers-cannot-reach-internet-in-bridge-network-mode

Compare host and bridge network settings. I don't know what these are supposed to look like.

```bash
docker network inspect host bridge | jq -c '.[]' | qsv jsonl | qsv transpose | csvlook
```

| Name               | host                                                             | bridge                                                           |
|--------------------|------------------------------------------------------------------|------------------------------------------------------------------|
| Id                 | 3df152a6d6cfcccb7d801ee15d7ccad710786e2872bf9f23833a0404471b589b | 459c38e7e3d2244a3c9b0f6f6e34ad6996f4b5be84051d0bac8636235be3ecb6 |
| Created            | 2021-01-07T22:19:59.311602644+01:00                              | 2024-06-20T12:38:42.195740682+02:00                              |
| Scope              | local                                                            | local                                                            |
| Driver             | host                                                             | bridge                                                           |
| EnableIPv6         | false                                                            | false                                                            |
| IPAM.Driver        | default                                                          | default                                                          |
| IPAM.Options       |                                                                  |                                                                  |
| IPAM.Config        |                                                                  | {"Subnet":"172.17.0.0/16","Gateway":"172.17.0.1"}                |
| Internal           | false                                                            | false                                                            |
| Attachable         | false                                                            | false                                                            |
| Ingress            | false                                                            | false                                                            |
| ConfigFrom.Network |                                                                  |                                                                  |
| ConfigOnly         | false                                                            | false                                                            |

```
$ ifconfig
docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        inet6 fe80::42:19ff:fe74:b1b  prefixlen 64  scopeid 0x20<link>
        ether 02:42:19:74:0b:1b  txqueuelen 0  (Ethernet)
        RX packets 45  bytes 2420 (2.4 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 40  bytes 5800 (5.8 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

```
$ networkctl
WARNING: systemd-networkd is not running, output will be incomplete.

IDX LINK            TYPE     OPERATIONAL SETUP    
  1 lo              loopback n/a         unmanaged
  2 enp0s31f6       ether    n/a         unmanaged
  3 wlp61s0         wlan     n/a         unmanaged
  4 wwan0           wwan     n/a         unmanaged
  5 br-243781db4596 bridge   n/a         unmanaged
  6 br-34ec3a34ef93 bridge   n/a         unmanaged
  7 br-3ff5d0a73058 bridge   n/a         unmanaged
  8 docker0         bridge   n/a         unmanaged
  9 br-9aee5e3f4add bridge   n/a         unmanaged
 10 br-a06628d0760f bridge   n/a         unmanaged
 11 br-eee029a4ec31 bridge   n/a         unmanaged

11 links listed.
```

```
$ ip route
default via 192.168.186.248 dev wlp61s0 proto dhcp metric 600 
169.254.0.0/16 dev wlp61s0 scope link metric 1000 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
172.18.0.0/16 dev br-eee029a4ec31 proto kernel scope link src 172.18.0.1 linkdown 
172.19.0.0/16 dev br-a06628d0760f proto kernel scope link src 172.19.0.1 linkdown 
172.20.0.0/16 dev br-34ec3a34ef93 proto kernel scope link src 172.20.0.1 linkdown 
172.21.0.0/16 dev br-3ff5d0a73058 proto kernel scope link src 172.21.0.1 linkdown 
172.22.0.0/16 dev br-9aee5e3f4add proto kernel scope link src 172.22.0.1 linkdown 
192.168.16.0/20 dev br-243781db4596 proto kernel scope link src 192.168.16.1 linkdown 
192.168.186.0/24 dev wlp61s0 proto kernel scope link src 192.168.186.172 metric 600 
```

The fix there was to delete a narrower route that badly routed the packets. I don't have one.

Try DFA Docker.

```
$ docker system prune -a --volumes
WARNING! This will remove:
  - all stopped containers
  - all networks not used by at least one container
  - all anonymous volumes not used by at least one container
  - all images without at least one container associated to them
  - all build cache
...
Total reclaimed space: 23.83GB
```

No change.

```
$ docker run --rm busybox ping google.com
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
ec562eabd705: Pull complete 
Digest: sha256:9ae97d36d26566ff84e8893c64a6dc4fe8ca6d1144bf5b87b2b85a32def253c7
Status: Downloaded newer image for busybox:latest
ping: bad address 'google.com'
```

Uninstall Docker.

```bash
sudo apt-get remove docker-ce
```

Restart.

Reinstall Docker.

```bash
sudo apt-get install docker-ce
```

That appears to have fixed the networking problem.

The DNS command returns the expected response immediately.

```
$ docker run --rm busybox nslookup google.com
Server:		192.168.1.1
Address:	192.168.1.1:53

Non-authoritative answer:
Name:	google.com
Address: 142.250.187.206

Non-authoritative answer:
Name:	google.com
Address: 2a00:1450:4009:818::200e
```
